cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/store/DefaultMetadataStore.java

```diff
@@ -720,6 +737,30 @@ public boolean isUpgradeRequired(MetadataScope scope) {
     return false;
   }
 
+  /**
+   * Scan the metadata store with a query which matches the index generated by
+   * {@link NearestKnownAncestorIndexer} to list all the sub entities which are associated with the given
+   * metadataEntity if it is a custom {@link MetadataEntity}.
+   * @param metadataEntity the {@link MetadataEntity} whose sub entities needs to be found
+   * @return a {@link Set} of {@link MetadataSearchResultRecordV2} which contains all the sub entities for the given
+   * metadataEntity if it is a custom entity else an empty set.
+   */
+  private Set<MetadataSearchResultRecordV2> getSubEntities(MetadataEntity metadataEntity) {
+    Set<MetadataSearchResultRecordV2> subEntities = new HashSet<>();
+    EntityId selfOrParentEntityId = EntityId.getNearestKnownEntity(metadataEntity);
+    // if a known entity is being deleted then we need to delete all it's sub-entity metadata too
+    if (selfOrParentEntityId.getEntityType().toString().equalsIgnoreCase(metadataEntity.getType())) {
```

Comment:
this is another place that would benefit from moving the logic into MetadataEntity